File: solutions_2453486_1/Python/XYZ791/tic.py

def processCase(data):
	# do some counting
	rows_x = [0,0,0,0]
	cols_x = [0,0,0,0]
	rows_o = [0,0,0,0]
	cols_o = [0,0,0,0]
	
	incomplete = False
	
	for row in range(4):
		for col in range(4):
			value = data[row][col]
			if value == 'X':
				rows_x[row] = rows_x[row] + 1
				cols_x[col] = cols_x[col] + 1
			elif value == 'O':
				rows_o[row] = rows_o[row] + 1
				cols_o[col] = cols_o[col] + 1
			elif value == 'T':
				rows_x[row] = rows_x[row] + 1
				cols_x[col] = cols_x[col] + 1
				rows_o[row] = rows_o[row] + 1
				cols_o[col] = cols_o[col] + 1
			elif value == '.':
				incomplete = True
		
		# check current row
		if rows_x[row] >= 4:
			return "X won"
		elif rows_o[row] >= 4:
			return "O won"
	
	# check cols
	for col in range(4):
		if cols_x[col] >= 4:
			return "X won"
		elif cols_o[col] >= 4:
			return "O won"
	
	# check leading diagonal
	diag1_x = 0
	diag1_o = 0
	for counter in range(4):
		value = data[counter][counter]
		if value == 'X':
			diag1_x = diag1_x + 1
		elif value == 'O':
			diag1_o = diag1_o + 1
		elif value == 'T':
			diag1_x = diag1_x + 1
			diag1_o = diag1_o + 1
	if diag1_x >= 4:
		return "X won"
	elif diag1_o >= 4:
		return "O won"
	
	# check other diagonal
	diag2_x = 0
	diag2_o = 0
	for counter in range(4):
		value = data[counter][3-counter]
		if value == 'X':
			diag2_x = diag2_x + 1
		elif value == 'O':
			diag2_o = diag2_o + 1
		elif value == 'T':
			diag2_x = diag2_x + 1
			diag2_o = diag2_o + 1
	if diag2_x >= 4:
		return "X won"
	elif diag2_o >= 4:
		return "O won"
	
	if incomplete:
		return "Game has not completed"
	else:
		return "Draw"

# open input file
in_file = open("in", "r")
n_cases = int(in_file.readline())

# open output file
out_file = open("out", "w")

# read then process test cases
for test_case in range(n_cases):
	case = []
	for line in range(4):
		case.append(in_file.readline())
	in_file.readline()
	
	out_file.write("Case #{}: {}\n".format(test_case+1, processCase(case)))

in_file.close()
out_file.close()

File: solutions_2453486_1/C++/timics23/tic.cpp
#include<stdio.h>
int T,a[5][5],ok=0,done=0;
char c;
int main()
{
    freopen("input.in","r",stdin);
    freopen("bla.out","w",stdout);
    scanf("%d",&T);
    for(int k=1;k<=T;++k)
    {
        scanf("%c",&c);
        ok=0;
        done=0;
        for(int i=1;i<=4;++i)
        {
            for(int j=1;j<=4;++j)
            {
                a[i][j]=0;
                scanf("%c",&c);
                if(c=='X')
                    a[i][j]=1;
                else if(c=='O')
                    a[i][j]=-1;
                else if(c=='T')
                    a[i][j]=8;
                else
                    done=1;

            }
            scanf("%c",&c);
        }
        int x=a[1][1]+a[1][2]+a[1][3]+a[1][4];

        if(x==4 || x==11)
        {
            if(ok==0)
                printf("Case #%d: X won\n",k);
            ok=1;
        }
        if(x==-4 || x==5)
        {
            if(ok==0)
            printf("Case #%d: O won\n",k);
            ok=1;
        }

        x=a[2][1]+a[2][2]+a[2][3]+a[2][4];

        if(x==4 || x==11)
        {
            if(ok==0)
            printf("Case #%d: X won\n",k);
            ok=1;
        }
        if(x==-4 || x==5)
        {
            if(ok==0)
            printf("Case #%d: O won\n",k);
            ok=1;
        }

        x=a[3][1]+a[3][2]+a[3][3]+a[3][4];

        if(x==4 || x==11)
        {
            if(ok==0)
            printf("Case #%d: X won\n",k);
            ok=1;
        }
        if(x==-4 || x==5)
        {
            if(ok==0)
            printf("Case #%d: O won\n",k);
            ok=1;
        }

        x=a[4][1]+a[4][2]+a[4][3]+a[4][4];

        if(x==4 || x==11)
        {
            if(ok==0)
            printf("Case #%d: X won\n",k);
            ok=1;
        }
        if(x==-4 || x==5)
        {
            if(ok==0)
            printf("Case #%d: O won\n",k);
            ok=1;
        }

        x=a[1][1]+a[2][1]+a[3][1]+a[4][1];

        if(x==4 || x==11)
        {
            if(ok==0)
            printf("Case #%d: X won\n",k);
            ok=1;
        }
        if(x==-4 || x==5)
        {
            if(ok==0)
            printf("Case #%d: O won\n",k);
            ok=1;
        }

        x=a[1][4]+a[2][4]+a[3][4]+a[4][4];

        if(x==4 || x==11)
        {
            if(ok==0)
            printf("Case #%d: X won\n",k);
            ok=1;
        }
        if(x==-4 || x==5)
        {
            if(ok==0)
            printf("Case #%d: O won\n",k);
            ok=1;
        }

        x=a[1][2]+a[2][2]+a[3][2]+a[4][2];

        if(x==4 || x==11)
        {
            if(ok==0)
            printf("Case #%d: X won\n",k);
            ok=1;
        }
        if(x==-4 || x==5)
        {
            if(ok==0)
            printf("Case #%d: O won\n",k);
            ok=1;
        }

        x=a[1][3]+a[2][3]+a[3][3]+a[4][3];

        if(x==4 || x==11)
        {
            if(ok==0)
            printf("Case #%d: X won\n",k);
            ok=1;
        }
        if(x==-4 || x==5)
        {
            if(ok==0)
            printf("Case #%d: O won\n",k);
            ok=1;
        }

        x=a[1][1]+a[2][2]+a[3][3]+a[4][4];

        if(x==4 || x==11)
        {
            if(ok==0)
            printf("Case #%d: X won\n",k);
            ok=1;
        }
        if(x==-4 || x==5)
        {
            if(ok==0)
            printf("Case #%d: O won\n",k);
            ok=1;
        }

        x=a[1][4]+a[2][3]+a[3][2]+a[4][1];

        if(x==4 || x==11)
        {
            if(ok==0)
            printf("Case #%d: X won\n",k);
            ok=1;
        }
        if(x==-4 || x==5)
        {
            if(ok==0)
            printf("Case #%d: O won\n",k);
            ok=1;
        }

        if(ok==0)
        {
            if(done==0)
                printf("Case #%d: Draw\n",k);
            else
                printf("Case #%d: Game has not completed\n",k);
        }

    }
    return 0;
}

